File: implementations/javascript/merge_sort.js
const SIZES = { small: 100, medium: 10000, large: 1000000 };
const n = SIZES[process.argv[2]];

function merge(arr, temp, left, mid, right) {
    for (let i = left; i <= right; i++) temp[i] = arr[i];
    let i = left, j = mid + 1, k = left;
    while (i <= mid && j <= right)
        arr[k++] = temp[i] <= temp[j] ? temp[i++] : temp[j++];
    while (i <= mid) arr[k++] = temp[i++];
    while (j <= right) arr[k++] = temp[j++];
}

function mergeSortImpl(arr, temp, left, right) {
    if (left >= right) return;
    const mid = left + Math.floor((right - left) / 2);
    mergeSortImpl(arr, temp, left, mid);
    mergeSortImpl(arr, temp, mid + 1, right);
    merge(arr, temp, left, mid, right);
}

const arr = new Array(n);
for (let i = 0; i < n; i++) arr[i] = n - i;
const temp = new Array(n);
if (n > 1) mergeSortImpl(arr, temp, 0, n - 1);
console.log(arr[n - 1]);

File: implementations/python/merge_sort.py
import sys

SMALL  = 100
MEDIUM = 10000
LARGE  = 1000000

SIZES = {"small": SMALL, "medium": MEDIUM, "large": LARGE}

def merge(arr, temp, left, mid, right):
    for i in range(left, right + 1):
        temp[i] = arr[i]
    i, j, k = left, mid + 1, left
    while i <= mid and j <= right:
        if temp[i] <= temp[j]:
            arr[k] = temp[i]
            i += 1
        else:
            arr[k] = temp[j]
            j += 1
        k += 1
    while i <= mid:
        arr[k] = temp[i]
        i += 1
        k += 1
    while j <= right:
        arr[k] = temp[j]
        j += 1
        k += 1

def merge_sort_impl(arr, temp, left, right):
    if left >= right:
        return
    mid = left + (right - left) // 2
    merge_sort_impl(arr, temp, left, mid)
    merge_sort_impl(arr, temp, mid + 1, right)
    merge(arr, temp, left, mid, right)

if __name__ == "__main__":
    n = SIZES[sys.argv[1]]
    arr = [n - i for i in range(n)]
    temp = [0] * n
    if n > 1:
        merge_sort_impl(arr, temp, 0, n - 1)
    print(arr[n - 1])
